### PR TITLE
[MRG] Fix scoring

### LINKED ISSUE #63: Usage of scorer is incorrect
We use `scorer._score_func(y_train, predicted_train)` where `predicted_train = self._predict_binned(X_binned_train)`.

`predict_binned` outputs classes for classification. This won't work for e.g. `scoring='neg_log_loss'` which requires probabilities from `predict_proba()`.

One solution is to get rid of `scoring` and check early stopping based on the loss.

If we want to keep `scoring`, I think we'll need to always use `scorer.__call__()` instead of `scorer._score_func()`, so that we don't reinvent the wheel.

The signature of `scorer.__call__()` is `estimator, X, y`, so we'll need to pass `self`. I think this is doable. I see 2 options:
- We pass `X_binned_train` to `scorer.__call__()` and `predict()` (and `predict_proba`) has to figure out whether its input `X` is pre-binned or not, e.g. by checking that the dtype is `uint8`. This forces users to never use `uint8` arrays.
- We pass `X_train` instead of `X_binned_train`. This is slightly slower.

In both cases it's better to have a C-contiguous `X` for predictions. It's faster than with Fortran arrays. Here are 3 runs of `bench_predictors.py` from #62 (5e6 samples)

 Pre-binned F  .5456s
 Pre-binned C  .5287s 
 numerical F   .7616s 
 numerical C   .5716s 

Pre-binned F .4914s
Pre-binned C .4690s
numerical F .6745s
numerical C .5033s

Pre-binned F .5019s
Pre-binned C .4724s
numerical F .6788s
numerical C .5099s

### PR DESCRIPTION
**Pull request by NicolasHug**
Fixes #63 

I went for the first solution: `_raw_predict` considers that `X` is pre-binned iff `X.dtype == np.uint8`, and we can use `scorer.__call__()` naturally.

In `fit()`,  when checking for early stopping, `_raw_predict()` will be called on either `X_binned_train` or `X_binned_val`. Both are F-array. Do we want to have copies `X_binned_train_C` and `X_binned_val_C`?
I fear that duplicating the training set is quite memory intensive.

### CODE REVIEW

**Comment by ogrisel** [comment]
> I fear that duplicating the training set is quite memory intensive.

You can do a quick benchmark to check the performance impact by enabling the copy in a branch.
But I share your concern with the memory usage. Let's keep it as it (no copy) if the quick benchmark does not show a big improvement in the full fit time of a non-trivial dataset (e.g. Higgs).

**Comment by NicolasHug** [comment]
I completely missed this but:

- We currently **are** copying the training set and converting it to a C array, if it's bigger than 10000 samples. So we might just as well *copy and convert* all the time, since this copy is of at most 10000 samples. Should we make `subsample_size` a parameter?
- We can always convert `X_binned_val` to C since we don't need to copy it, so no memory issue here. 

I made a few experiments with conversion to C or not and it's a bit faster, probably worth it:

On Higgsboson 1e6 samples, average time spent in `check_early_stopping` is .340s with F arrays and .097 with C arrays.

File: pygbm/gradient_boosting.py [hunk 1]
```diff
@@ -135,22 +135,24 @@ def fit(self, X, y):
                     f'{self.validation_split}. Use more training data or '
                     f'adjust validation_split.'
                 )
-            # Histogram computation is faster on feature-aligned data.
+            # Predicting is faster of C-contiguous arrays, training is faster
+            # on Fortran arrays.
```
Comment (ogrisel) [inline pygbm/gradient_boosting.py]: Good comment.

File: pygbm/gradient_boosting.py [hunk 2]
```diff
@@ -293,25 +291,29 @@ def _should_stop(scores):
...
+            # negative. We thus need to multiply tol by -1 in those cases,
+            # else we would actually decrease prev_score instead of increasing
+            # it.
+
             current_score = scores[-1]  # score at current iteration
             previous_scores = scores[-self.n_iter_no_change:-1]
+            sign = self.scorer_._sign
             return all(
-                current_score < prev_score * (1 + self.tol * scorer._sign)
+                current_score < prev_score * (1 + self.tol * sign)
```
[the remark this answers is not in the document]
Comment (ogrisel) [reply pygbm/gradient_boosting.py]: Actually, the previous code is not the stopping criterion I want. May I suggest the following:


```python
reference_position = self.n_iter_no_change + 1
if len(scores) < reference_position:
    reference_score = scores[-reference_position]
    recent_scores = scores[-reference_position + 1:]
    recent_improvements = [score > reference_score * (1 - tol)
                           for score in recent_scores]
    return not any(recent_improvements)
```

[the remark this answers is not in the document]
Comment (ogrisel) [reply pygbm/gradient_boosting.py]: Thinking about it, I think we should use an absolute value for `tol`: it's not intuitive interpret it as a fraction of the reference score value as it will evolve over time.

Therefore I suggest the following instead:

```python
reference_position = self.n_iter_no_change + 1
if len(scores) < reference_position:
    reference_score = scores[-reference_position] - tol  # higher is always better
    recent_scores = scores[-reference_position + 1:]
    recent_improvements = [score > reference_score
                           for score in recent_scores]
    return not any(recent_improvements)
```

[the remark this answers is not in the document]
Comment (NicolasHug) [reply pygbm/gradient_boosting.py]: Ok I'll use an absolute tol, I think that's what sklearn is doing as well, and we don't have to worry about the sign anymore.

------

Let me know if I made a mistake in the previous version but I think that 

```python
all(current_score < prev_score * (1 + tol) for prev_score in previous_scores)
```

is incorrect when `scorer._sign` is `-1`, because in this case `prev_score` is negative (the closer it is to 0, the better). Multiplying it by `(1 + tol)` will decrease it instead of increasing it. Hence the use of  `1 + tol * sign`.

[the remark this answers is not in the document]
Comment (NicolasHug) [reply pygbm/gradient_boosting.py]: Did you mean

```python
reference_score = scores[-reference_position] + tol
```

with a plus?

My understanding is taht the higher the tol, the more likely we are to early stop.

[the remark this answers is not in the document]
Comment (ogrisel) [reply pygbm/gradient_boosting.py]: > The best I have so far is: ...

I like this plain English explanation.

> My understanding is that before (and in sklearn), the higher the tol, the more likely we are to early stop. 

Yes I agree. My code is wrong. This is indeed very hard to reason correctly about this thing. Here is my "final" suggestion:


```python
reference_position = self.n_iter_no_change + 1
if len(scores) < reference_position:
    # Remarks: a higher score is always better
    # Higher tol means that it will be hard for subsequent iteration to be
    # considered an improvement upon the reference score and therefore
    # more likely to early stop because of the lack of significant improvement.
    reference_score = scores[-reference_position] + tol
    recent_scores = scores[-reference_position + 1:]
    recent_improvements = [score > reference_score
                           for score in recent_scores]
    return not any(recent_improvements)
```

[the remark this answers is not in the document]
Comment (ogrisel) [reply pygbm/gradient_boosting.py]: > Personally I found the previous strategy easier to explain and to understand.

I cannot understand the logic of the previous strategy. I prefer my latest suggestion.

[the remark this answers is not in the document]
Comment (ogrisel) [reply pygbm/gradient_boosting.py]: Could you please extract `_should_stop(scores, n_iter_no_change, tol)` as a standalone function (or method) so that we can write unittest for it?


```python
def test_shoud_stop():
    # not enough iterations
    assert not _should_stop([], n_iter_no_change=1, tol=0.001)

    assert not _should_stop([1, 1, 1], n_iter_no_change=5, tol=0.001)
    assert not _should_stop([1] * 5, n_iter_no_change=5, tol=0.001)

    # still making significant progress up to tol
    assert not _should_stop([1, 2, 3, 4, 5, 6], n_iter_no_change=5, tol=0.001)
    assert not _should_stop([1, 2, 3, 4, 5, 6], n_iter_no_change=5, tol=0.)
    assert not _should_stop([1, 2, 3, 4, 5, 6], n_iter_no_change=5, tol=0.999)

    # no significant progress according to tol
    assert _should_stop([1] * 6, n_iter_no_change=5, tol=0.)
    assert _should_stop([1] * 6, n_iter_no_change=5, tol=0.001)
    assert _should_stop([1, 2, 3, 4, 5, 6], n_iter_no_change=5, tol=2)

```

File: tests/test_gradient_boosting.py [hunk 3]
```diff
@@ -166,6 +166,32 @@ def test_early_stopping_classification(data, scoring, validation_split, tol):
...
+
+    assert not should_stop([1, 1, 1], n_iter_no_change=5, tol=0.001)
+    assert not should_stop([1] * 5, n_iter_no_change=5, tol=0.001)
+
+    # still making significant progress up to tol
+    assert not should_stop([1, 2, 3, 4, 5, 6], n_iter_no_change=5, tol=0.001)
+    assert not should_stop([1, 2, 3, 4, 5, 6], n_iter_no_change=5, tol=0.)
+    assert not should_stop([1, 2, 3, 4, 5, 6], n_iter_no_change=5, tol=0.999)
+    assert not should_stop([1, 2, 3, 4, 5, 6], n_iter_no_change=5,
+                           tol=5 - 1e-5)
```
Comment (ogrisel) [inline tests/test_gradient_boosting.py]: Do you mean `1 - 1e-5`?

Comment (ogrisel) [reply tests/test_gradient_boosting.py]: Hum you are right, it's me that was wrong with my 0.999 tol...

Comment (NicolasHug) [reply tests/test_gradient_boosting.py]: Also I changed 

```
    assert _should_stop([1, 2, 3, 4, 5, 6], n_iter_no_change=5, tol=2)
```

to

```
    assert _should_stop([1, 2, 3, 4, 5, 6], n_iter_no_change=5, tol=5)
```
